runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/ProcessRemoteBundleOperation.java

```diff
@@ -48,25 +62,119 @@
+      DataflowOperationContext operationContext,
       StageBundleFactory stageBundleFactory,
       Map<String, OutputReceiver> outputReceiverMap) {
-    super(EMPTY_RECEIVER_ARRAY, context);
+    super(EMPTY_RECEIVER_ARRAY, operationContext);
+
     this.stageBundleFactory = stageBundleFactory;
-    this.outputReceiverMap = outputReceiverMap;
     this.stateRequestHandler = StateRequestHandler.unsupported();
     this.progressHandler = BundleProgressHandler.ignored();
+    this.executionContext = executionContext;
+    this.executableStage = executableStage;
+    this.outputReceiverMap = outputReceiverMap;
+
+    this.timerOutputIdToSpecMap = new HashMap<>();
+    this.timerIdToKey = new HashMap<>();
+    this.timerIdToPayload = new HashMap<>();
+    this.timerIdToTimerSpecMap = new HashMap<>();
+
+    ProcessBundleDescriptors.ExecutableProcessBundleDescriptor executableProcessBundleDescriptor =
+        stageBundleFactory.getProcessBundleDescriptor();
+
+    BeamFnApi.ProcessBundleDescriptor processBundleDescriptor =
+        executableProcessBundleDescriptor.getProcessBundleDescriptor();
+
+    // Create and cache lookups so that we don't have to dive into the ProcessBundleDescriptor
+    // later.
+    fillTimerSpecMaps(
+        executableProcessBundleDescriptor, this.timerIdToTimerSpecMap, this.timerOutputIdToSpecMap);
+    this.timerWindowCodersMap =
+        fillTimerWindowCodersMap(
+            processBundleDescriptor, this.timerIdToTimerSpecMap, this.executableStage);
+  }
+
+  // Fills the given maps from timer the TimerSpec definitions in the process bundle descriptor.
+  private void fillTimerSpecMaps(
+      ProcessBundleDescriptors.ExecutableProcessBundleDescriptor processBundleDescriptor,
+      Map<String, ProcessBundleDescriptors.TimerSpec> timerIdToTimerSpecMap,
+      Map<String, ProcessBundleDescriptors.TimerSpec> timerOutputIdToSpecMap) {
+    processBundleDescriptor
+        .getTimerSpecs()
+        .values()
+        .forEach(
+            transformTimerMap -> {
+              for (ProcessBundleDescriptors.TimerSpec timerSpec : transformTimerMap.values()) {
+                timerIdToTimerSpecMap.put(timerSpec.timerId(), timerSpec);
+                timerOutputIdToSpecMap.put(timerSpec.outputCollectionId(), timerSpec);
+              }
+            });
+  }
+
+  // Retrieves all window coders for all TimerSpecs.
+  private Map<String, Coder<BoundedWindow>> fillTimerWindowCodersMap(
+      BeamFnApi.ProcessBundleDescriptor processBundleDescriptor,
+      Map<String, ProcessBundleDescriptors.TimerSpec> timerIdToTimerSpecMap,
+      ExecutableStage executableStage) {
+    Map<String, Coder<BoundedWindow>> timerWindowCodersMap = new HashMap<>();
+    for (RunnerApi.PTransform pTransform : processBundleDescriptor.getTransformsMap().values()) {
+      for (String timerId : timerIdToTimerSpecMap.keySet()) {
+        if (!pTransform.getInputsMap().containsKey(timerId)) {
+          continue;
```

Comment:
Why is it OK to continue?

Reply:
Because not every PTransform has a timer. This is just a quick thing to find the matching timer.